src/Bicep.LangServer/Registry/ModuleRestoreScheduler.cs

```diff
@@ -0,0 +1,134 @@
+
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+
+        // block on initial wait until signaled
+        private readonly ManualResetEventSlim manualResetEvent = new(false);
+
+        private bool disposed = false;
+        private Task? consumerTask;
+
+        public ModuleRestoreScheduler(IModuleDispatcher moduleDispatcher)
+        {
+            this.moduleDispatcher = moduleDispatcher;
+        }
+
+        /// <summary>
+        /// Requests that the specified modules be restored to the local file system.
+        /// Does not wait for the operation to complete and returns immediately.
+        /// </summary>
+        /// <param name="references">The module references</param>
+        public void RequestModuleRestore(ICompilationManager compilationManager, DocumentUri documentUri, IEnumerable<ModuleDeclarationSyntax> references)
+        {
+            this.CheckDisposed();
+            var item = new QueueItem(compilationManager, documentUri, references.ToImmutableArray());
+            lock (this.queue)
+            {
+                this.queue.Enqueue(item);
+
+                // notify consumer about new items
+                this.manualResetEvent.Set();
+            }
+        }
+
+        public void Start()
+        {
+            this.CheckDisposed();
+            this.consumerTask = Task.Factory.StartNew(this.ProcessQueueItems, this.cancellationTokenSource.Token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+        }
+
+        public void Dispose()
+        {
+            // this is a sealed class - no need for full IDisposable implementation
+            if(!this.disposed)
+            {
+                this.disposed = true;
+                if (this.consumerTask is not null)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    this.consumerTask = null;
+                }
+            }
+        }
+
+        private void ProcessQueueItems()
+        {
+            var token = this.cancellationTokenSource.Token;
+            while (true)
+            {
+                this.manualResetEvent.Wait(token);
+
+                var notifications = new HashSet<CompletionNotification>();
+                var references = new List<ModuleDeclarationSyntax>();
+                lock (this.queue)
+                {
+                    this.UnsafeCollectModuleReferences(notifications, references);
+                    Debug.Assert(this.queue.Count == 0, "this.queue.Count == 0");
+
+                    // queue has been consumed - next iteration should block until more items have been added
+                    this.manualResetEvent.Reset();
+                }
+
+                // this blocks until restore is completed
+                // the dispatcher stores the results internally and manages their lifecycle
+                token.ThrowIfCancellationRequested();
+                if(!this.moduleDispatcher.RestoreModules(references))
```

Comment:
Should we embrace `async` here? Maybe not an easy refactor, but feels like it's going to be a necessity at some point.

Reply:
One benefit I can think of BTW is that we'll honor cancellation token e.g. if a particular module restore is taking a very long time, and the server is shut down

Reply:
Yeah that's the plan for a later PR. (I will have to integrate an SDK to perform module restore, which will require refactoring to `async`.)